src/Psl/Shell/execute.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Shell;
+
+use Psl\Dict;
+use Psl\Env;
+use Psl\Str;
+use Psl\Vec;
+
+use function fclose;
+use function is_dir;
+use function is_resource;
+use function proc_close;
+use function proc_open;
+use function stream_get_contents;
+
+/**
+ * Execute an external program.
+ *
+ * @param string                $command            The command to execute.
+ * @param list<string>          $arguments          The command arguments listed as separate entries.
+ * @param string                $working_directory  The initial working directory for the command.
+ *                                                  This must be an absolute directory path, or null
+ *                                                  if you want to use the default value (the working dir of the
+ *                                                  current directory )
+ * @param array<string, string> $environment        An array with the environment variables for the command that
+ *                                                  will be run.
+ * @param bool                  $escape_arguments   If set to true ( default ), all $arguments will be escaped using
+ *                                                  `Shell\escape_argument`.
+ *
+ * @psalm-taint-sink shell $command
+ *
+ * @throws Exception\FailedExecutionException   In case the command resulted in an exit code other than 0.
+ * @throws Exception\PossibleAttackException    In case the command being run is suspicious ( e.g: contains NULL byte ).
+ * @throws Exception\RuntimeException           In case $working_directory doesn't exist, or unable to create a new
+ *                                              process.
+ */
+function execute(
+    string $command,
```

Comment:
The command is actually also a cli argument.
In PHP, they always keep this separated. But what is the purpose of keeping those 2 things separated?
They are converted to a string command anyways.

Reply:
because in most cases, you already know the command beforehand, but likely not the arguments:

```php
$tmp = Env\temp_dir() . '/bar';

Shell\execute('foo', ['-f', $temp]);

return File\read($temp);
```

This also allows us to escape the arguments individually.

Reply:
You already have that choice though: You can either use the flag `escape_arguments` or you don't.
Even though you know the command in advance, you can still just pass it as a first argument of the array.
I'm not too sure that having the main command separately really adds any benefits

Reply:
assuming we merge `$command` with `$arguments`, `$command` will also be escaped as an argument, but if you don't want that, you will have to set `$escape_arguments` to `false`, and escape arguments yourself before calling `Shell\execute` in order to leave the command as is.

Reply:
In which case wouldn't you want the command itself to be escaped?
Are you ever providing a command escaped directly as input?

Most of the times, you just pass the commandname. Like E.g: run  `php` as command, then it wraps it with quotes like this `"php"`. Which is basically what symfony/process does as well.

Reply:
in this case, the command is escaped - or known to be safe ( `@psalm-taint-sink shell $command` ), psalm taint analysis will catch it in case it's something like `Shell\execute($_POST['command'])`.

and another thing: `Shell\execute('foo');` is better/easier to type than `Shell\execute(['foo']);`

> Which is basically what symfony/process does as well.

you shouldn't really compare this to `symfony/process`, but rather to `exec`/`shell_exec`